src/platform/nodejs/persist/fileSystemPersist.ts

```diff
@@ -0,0 +1,132 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import { Logger } from '@opentelemetry/api';
+import { NoopLogger } from '@opentelemetry/core';
+import { PersistentStorage } from '../../../types';
+import { AzureExporterConfig, DEFAULT_EXPORTER_CONFIG } from '../../../config';
+import { confirmDirExists, getShallowDirectorySize } from './fileSystemHelpers';
+
+export class FileSystemPersist implements PersistentStorage {
+  static TEMPDIR_PREFIX = 'ot-azure-exporter-';
+
+  maxBytesOnDisk: number = 50_000_000; // ~50MB
+
+  private readonly _options: AzureExporterConfig;
+
+  private readonly _logger: Logger;
+
+  constructor(options: Partial<AzureExporterConfig> = {}) {
+    this._options = { ...DEFAULT_EXPORTER_CONFIG, ...options };
+    this._logger = options.logger || new NoopLogger();
+    if (!this._options.instrumentationKey) {
+      this._logger.error(
+        `No instrumentation key was provided to FileSystemPersister. Files may not be properly persisted`,
+      );
+    }
+  }
+
+  push(value: unknown[], cb: (err: Error | null, result?: boolean | undefined) => void): void {
+    this._logger.info('Pushing value to persistent storage', value.toString());
+    this._storeToDisk(JSON.stringify(value), cb);
+  }
+
+  shift(cb: (err: Error | null, value?: unknown[]) => void): void {
+    this._logger.info('Returning first member of filesystem');
+    this._getFirstFileOnDisk((error, buffer) => {
+      if (error) {
+        cb(error);
+      } else if (buffer) {
+        cb(null, JSON.parse(buffer.toString('utf8')));
+      } else {
+        cb(null);
+      }
+    });
+  }
+
+  /**
+   * Check for temp telemetry files
+   * reads the first file if exist, deletes it and tries to send its load
+   */
+  private _getFirstFileOnDisk(callback: (error: Error | null, value?: Buffer) => void): void {
+    const tempDir = path.join(
+      os.tmpdir(),
+      FileSystemPersist.TEMPDIR_PREFIX + this._options.instrumentationKey,
+    );
+
+    fs.stat(tempDir, (statErr: Error | null, stats: fs.Stats) => {
+      if (stats.isDirectory()) {
```

Comment:
Python implementation has mechanisms in place to prevent read/writing and sending a file that is being used in another process. I know JS does not have multiple threads, but is multi process scenarios possible?

Reply:
Good point. I don't see this happening unless the user tries to run nodejs twice on the same box sending to the same instrumentation key. Not sure when this would even be feasible. Anything containerized will use isolated filesystems. Regardless, the current implementation of delete --> send solves this.

Reply:
It could be possible having multiple nodejs instances for a system that is running multiple workers. I don't think it is only the double telemetry case that needs to be handled for concurrency. Consider the case of one worker trying to write to a file, but another trying to read/delete it at the same time. Some sort of synchronization mechanism is needed if multiple sources will be touching a critical section. Python implementation uses [leasing](https://github.com/microsoft/opentelemetry-azure-monitor-python/blob/master/azure_monitor/src/azure_monitor/storage.py#L146) of files for this. Could be an issue for a different PR though.

Reply:
Got it. Best practices is to run 1 nodejs instance per box/container. Running multiple instances on the same box, different ports would probably expose all sorts of other issues outside of this exporter. Either way, the file will not be readable and will error out here if any sort of syncing issues occur.